Webkit browser

TAKeanice opened the pull request:
Hello dears,

i made some progress on the webkit implementation of our browser, and I believe it is mostly done. There are a few things to clean up though, like the ability to change text sizes in browser tabs and some more todos.

In the long run, we could also replace the article view, but that one is more customized than the regular browser part.

Besides the todos that I already noted in the code, could you check out this branch and make a list of things you are missing?

TAKeanice commented:
The lack of printing ability could be solved this way:
https://github.com/marc-medley/WKWebView-Printing-Puzzle

Text resizing is an open "research" area:
https://stackoverflow.com/questions/34130319/how-to-de-and-increase-the-font-size-with-wkwebview
https://stackoverflow.com/questions/62525413/increasing-and-decreasing-text-size-of-wkwebview-on-macos?noredirect=1#comment111648311_62525413

barijaona commented:
Very interesting! Congratulations for all the work done!

But reviewing this pull request is more difficult than it should be… You are currently packing into a single PR many changes which do not have quite direct relationship.
   
I suggest you [hide some of the sausage making](http://sethrobertson.github.io/GitBestPractices/#sausage) :smile:
After a quick read of your changes, I think you can separate things into different pull requests : 
- code cleanup after the upgrade of our deployment target to macOS 10.11
- pull request of TAKeanice/MMTabBarView into MMTabBarView
- update of Vienna dependencies (MMTabBarView and Sparkle)
- replacement of WebView with WkWebView
- addition of tab order preferences (work in progress)

This can be done through extensive use of `git cherry-pick` and `git commit --amend` or with a GUI tool like [Gitup](https://github.com/git-up/GitUp)

While you are at it, you could adopt a more consistent style for your commit messages. Currently, you mix present, past and progressive tenses (`restore`, `restored` and `restoring`). It is recommended to use [present-tense, imperative-style](https://github.com/ViennaRSS/vienna-rss/blob/master/CONTRIBUTING.md#commit-messages).

The feature I miss most is the "RSS" button which is very handy for detecting a feed and subscribing to it…

Keep up the good work!

Eitot commented:
I am not sure if I have time this weekend, but I’ll try to. This PR is sorely needed though, so I am not too picky about it being large as long as it is somewhat self-contained. Maybe it could be merged onto a feature branch first and then cherry-picked onto master?

barijaona commented:
The tab bar needs improvement : especially, redirections should lead to updates of the displayed address.

TAKeanice commented:
Thanks for the two bugs reported @barijaona . I will look into both and fix the issues.
You are right about the separation,
- I can factor out the tab reading order preferences
- and create a pull request for MMTabBar

However, since the .xcodeproj xml is an entangled mess, I would like to combine
- the replacement of WKWebView
- upgrade of the Vienna dependencies
- and the code cleanup after the upgrade of our deployment target to macOS 10.11

because some of the changes are interdependent.

I would like to keep the content of the commits in general, since it is easier to resolve conflicts during a merge or rebase knowing what change was supposed to be in a commit. But I agree that I need to update the commit messages to adhere to the naming convention.

TAKeanice commented:
Here we go, I have made my sausage in the backroom and the commits above is the offer I have on my counter :) Though I accept any criticism that you return with after you grilled it thoroughly.

barijaona commented:
This crashes on launch here. In the log (edited for brievity);
````
Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value: file Vienna/Sources/Main window/ArticleController.swift, line 14
``````

josh64x2 commented:
Crashes for me too ☹️

TAKeanice commented:
> This crashes on launch here. In the log (edited for brievity);
> 
> ```
> Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value: file Vienna/Sources/Main window/ArticleController.swift, line 14
> ```
Strange, I cannot reproduce that at all. The referenced line does not force-unwrap anything.
Is it possibly related to my OS being BigSur Beta 20A5323l ? I am using Xcode 11.6

TAKeanice commented:
I also discovered that during my rebasing I missed the latest few commits. They are there now.

Eitot commented:
> Strange, I cannot reproduce that at all. The referenced line does not force-unwrap anything.
> Is it possibly related to my OS being BigSur Beta 20A5323l ? I am using Xcode 11.6

It seems that you are extending ArticleController, which has no nullability annotations. For some reason, the Swift compiler accepts `selectedArticle` accessor as a non-nil value and does not enforce unwrapping. Either that, or the problem is in the Apple frameworks.

Maybe the debugger will help to out what is nil here.

josh64x2 commented:
OK so this time when I launched Vienna, I didn't get the crash. But when I changed my selection in the sidebar, I did get the crash.
<img width="1114" alt="image" src="https://user-images.githubusercontent.com/799345/90765110-bd2b0200-e32c-11ea-990a-92b01ad5e953.png">

It's happening when trying to validate toolbar button items. This is the UI when the crash happened.
<img width="1092" alt="image" src="https://user-images.githubusercontent.com/799345/90765265-fb282600-e32c-11ea-8951-5e90b1f59285.png">

Big Sur public beta 20A5343j and Xcode 12b5

TAKeanice commented:
I think there is either some NS_ASSUME_NONNULL annotation or that is the implicit default. I will figure it out later today.

barijaona commented:
I confirm the same problem with 10.15.6 / Xcode 11.6

TAKeanice commented:
@barijaona  the RSS button is back!

However the delegate for responding to a click on the RSS button is not hooked up yet. The old browser knew a lot about the Vienna folder structure etc and directly triggered the subscription, I would like to avoid this for the new browser. I wrote a todo in `MainWindowController.swift` for that. Also, the logic for subscribing to a feed found in some tab can be taken from `handleRSSPage` in `BrowserPane.m`

I will not be able to work on Vienna for the coming two weeks, so feel free and have a look at that and the remaining Todos in the various `BrowserTab*.swift` files, if you want to help getting this done more quickly. My focus for implementing the new browser was that it has a clean interface and no tight coupling to anything else of Vienna for reusability wherever we (or anyone else) please(s).

barijaona commented:
> Something is probably wrong with tab closing : test closing a tab which has sound playing. The tab disappears, but the sound continues playing.

This problem is back : encountered it with https://www.bbc.com/news/av/election-us-2020-53923363/trump-s-speech-from-criteria-met-to-falsehoods-and-lies
I played the video in its entirety, restarted it, closed the tab : sound continues.

barijaona commented:
It's not a priority, but I much prefer the previous logic regarding tab order…

For the time being, what I miss most is opening links on a new tab.

TAKeanice commented:
> For the time being, what I miss most is opening links on a new tab.

I have not changed anything regarding the tab order beyond plain MMTabBar. Could you elaborate on how the previous version differed? I also have not restored my idea of tab closing order yet, but I will do that in a future PR once this one is incorporated into Vienna.

TAKeanice commented:
> This problem is back : encountered it with https://www.bbc.com/news/av/election-us-2020-53923363/trump-s-speech-from-criteria-met-to-falsehoods-and-lies
> I played the video in its entirety, restarted it, closed the tab : sound continues.

was related due to the way the tab was closed: closing by key combination cmd+w behaved different than clicking x. That is now fixed, in both scenarios the tab is now closed properly.

barijaona commented:
> I have not changed anything regarding the tab order beyond plain MMTabBar. Could you elaborate on how the previous version differed?

Usual behaviour : when I close the current tab, the adjacent tab on the **right** get selected.
This branch's behaviour : when I close the current tab, the adjacent tab on the **left** get selected

barijaona commented:
For consistency of Xcode project organisation, I think we should keep all .xib files inside the Interfaces group

barijaona commented:
I don't want to trigger any religious war 😸, but for new files, we were supposed to use 4 spaces instead of tabs for indentation…

TAKeanice commented:
I will solve both those consistency issues at the end of the development. I am not yet sure whether I should extract the browser to a separate project and make it work with Carthage, if we decide on doing that the consistency issues would not be relevant.

josh64x2 commented:
@TAKeanice We would like to drop Carthage and move to SPM so keep that in mind if you move the browser into its own thing 😃

barijaona commented:
> @TAKeanice We would like to drop Carthage and move to SPM so keep that in mind if you move the browser into its own thing 😃

Related : https://github.com/ViennaRSS/MMTabBarView/issues/18

TAKeanice commented:
> For the time being, what I miss most is opening links on a new tab.

implemented, without localization but functionality is there. I also added an "open in Safari" item which we can replace by "open in default browser" and add functionality later. What do you think about having both, the option to "open in new tab" and to "open in [new tab in] background"?

barijaona commented:
> What do you think about having both, the option to "open in new tab" and to "open in [new tab in] background"?

We already have in the General Preferences panel the option boxes for : 
- Opening new links in the background
- Opening links in external (default) browser

For consistency and concision, the first menu item should just respect these choices.

barijaona commented:
> For consistency and concision, the first menu item should just respect these choices.

In current Vienna, option-click punctually reverses the set preferences.

barijaona commented:
> > I have not changed anything regarding the tab order beyond plain MMTabBar. Could you elaborate on how the previous version differed? I also have not restored my idea of tab closing order yet, but I will do that in a future PR once this one is incorporated into Vienna.
> 
> Usual behaviour : when I close the current tab, the adjacent tab on the **right** get selected.
> This branch's behaviour : when I close the current tab, the adjacent tab on the **left** get selected

You probably need to implement at least some parts of the tab ordering thing. The routine which was present in current master's `Browser.m` is broken in your branch's `WebViewBrowser.m` (missing `-configureTabClosingBehavior`) and absent in `TabbedBrowserViewController.swift`

barijaona commented:
as a temporary workaround for my own use, I just added `self.showNextTab()` to `func tabView(_ aTabView: NSTabView, willClose tabViewItem: NSTabViewItem)` in  TabbedBrowserViewController.swift

TAKeanice commented:
I planned on implementing the full tab ordering functionality. Originally, if I recall correctly, the tab closing was prevented by returning false from a delegate method, and then we selected the next tab and afterwards closed the previous tab. That way we prevented loading a tab that was not shown afterwards. However, I first need to work on figuring out which item (e.g. which link url) a user right-clicked on, to complete the context menu functionality. I have implemented an approach locally but will only be able to finish it next weekend.

TAKeanice commented:
What do you think about implementing the “which tab to open after closing the current one” part as a delegate method of MMTabBar?

barijaona commented:
> What do you think about implementing the “which tab to open after closing the current one” part as a delegate method of MMTabBar?

It makes sense.
cc @MiMo42

TAKeanice commented:
@barijaona I just implemented the old tab closing behaviour on the basis of the new MMTabBarView delegate method. Now what is missing are localized labels and the order of points in the context menu for link opening. Also, to get the new browser out as quickly as possible, I will for now leave it integrated into Vienna, and maybe factor it out into a Swift framework later. Thus things like indentation and the location of xib files need to be corrected.

TAKeanice commented:
So I finalized my context menu efforts and also improved the RSS button a bit, localized texts and adapted the style to fit the remaining parts of Vienna / the code style guidelines in general. In my view, the new browser is now at least on par with the old one in terms of functionality. I would like you all to review it once more, and make two lists: things that you deem necessary for the very first version of Vienna that includes the new browser, and further improvements. I would love having this in the next official Vienna release!

TAKeanice commented:
Oh and for any small changes, e.g. typos and the like, I also would appreciate if you just fixed it right away and made PRs to my branch. If you find the time.

barijaona reviewed (commented):
This PR is already great, but in my opinion, we should port its evolutions to ArticleView before merging. Having such behavioral differences between the primary tab and other tabs is an annoyance.

I find the animation of load bar a bit too distracting. We need something subtler.

After a few days of use, the things I miss the most are : 
- hint of subsequent URL when hovering over a link
- "Open Page in _(preferred browser)_"

Keep up the good work !

josh64x2 commented:
I'm getting this static analysis error in `TabbedBrowserViewController.swift`:
<img width="770" alt="image" src="https://user-images.githubusercontent.com/799345/99142287-ab48c900-26a7-11eb-8011-d12f31ad0a2a.png">

I'm also getting this warning on lines 50 and 72 of `CustomWKWebView.swift`:
<img width="756" alt="image" src="https://user-images.githubusercontent.com/799345/99142292-c0255c80-26a7-11eb-9917-76eaa718d2f2.png">

josh64x2 commented on Vienna/Sources/Main window/TabbedBrowserViewController.swift:
```diff
@@ -0,0 +1,359 @@
+    }
+}
+
+@available(OSX 10.10, *)
```
Since our minimum deployment target is now 10.11, can this be changed?

josh64x2 commented on Vienna/Sources/Main window/ArticleCellView.m:
```diff
@@ -32,31 +32,52 @@ -(instancetype)initWithFrame:(NSRect)frameRect
-		SEL loadFinishedSelector = NSSelectorFromString(@"webViewLoadFinished:");
-		[[NSNotificationCenter defaultCenter] addObserver:[controller.browser primaryTabItemView] selector:loadFinishedSelector name:WebViewProgressFinishedNotification object:articleView];
+
+		if (@available(macOS 10.10, *)) {
```
10.11 is our new minimum

josh64x2 commented on Vienna/Sources/Main window/ArticleCellView.m:
```diff
@@ -32,31 +32,52 @@ -(instancetype)initWithFrame:(NSRect)frameRect
 	return self;
 }
 
+-(void)initializeWebKitArticleView API_AVAILABLE(macosx(10.10)) {
```
10.11 minimum again... 🙃

josh64x2 reviewed (commented):
Looks great, and I can't wait to merge!

I've pointed out a few places where we check for macOS 10.10 when our new minimum is 10.11, so we can probably drop these checks. There a several more, but I didn't want to point them all out when you can just do a search in Xcode.

What's the plan to address all of the `TODO` items? Are they all going to be addressed after the PR is merged or will you be working on them before merge?

TAKeanice commented:
Yes, the 10.10 checks and the explicit returns are there because this was started long before we dropped 10.9, so I will remove them soon (or accept any or that does this). Many of the todos are about the ArticleView, which is not yet ready. The others need to be sorted into "nice to have" and "required", which I haven't done yet. I know of at least one, the response to JS alerts and authentication requests you the WKUIDelegate, that we should probably implement.

TAKeanice commented:
I have now included a WebKit-based ArticleView for the first time. There are still issues with it: It does not display the style correctly and if you open WebKitArticleTab.swift there are many TODOs, not all of which are actually necessary. I appreciate, besides the refinement of the browser part, any help in resolving those TODOs.

barijaona commented:
I started to rebase this PR on current `master`, trying to maintain basic functionality from commit to commit.
Here are my results so far: https://github.com/barijaona/vienna-rss/tree/webkit_browser2

It's a slow process, because of the many conflicts to deal with on `project.pbxproj` and `MainWindowController.xib`, but it's very useful for my understanding of the changes. I also hope to avoid inadvertant regressions. I don't know if it can be useful to other people.

I uninstalled SwiftLint for this, because it encumbers my work.

barijaona commented:
I finished rebasing @TAKeanice 's excellent work on current master. See branch https://github.com/barijaona/vienna-rss/tree/webkit_browser2
This makes analysis much easier. This also avoids having parts of history that Xcode is unable to build.
I also slightly edited some commit messages for legibility.

Differences with current state of this PR are very limited, so I am confident that @TAKeanice can easily restart from my branch.

TAKeanice commented:
Alright, I will have a look and close this PR afterwards. Will you create a new PR and allow edits from maintainers such that I can continue the work?

barijaona commented:
You can resubmit it yourself. This is probably better as you'll probably be in position to make more changes as you are more familiar with it.

barijaona commented:
Replaced by #1407
